tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/mhlo_shape_calc_placement.cc

```diff
@@ -0,0 +1,539 @@
+    // is GPU or CPU
+    addMemcpyNodes();
+  };
+
+ private:
+  // Place the shape calculation subgraph to Host
+  void placeShapeCalcSubgraphOnHost();
+  // Place any mhlo ops that calculates I32 on Host
+  void placeI32OpsOnHost();
+  // Insert h2d and d2h OP on cross device edges.
+  void addMemcpyNodes();
+
+  // for rule based placement strategy, the placement of the op in the list
+  // is up to the placement of the dominant operand
+  const std::unordered_map<std::string, /*dominant operand index*/ int>
+      kPlaceRuleMap = {{"mhlo.dynamic_gather", /*operand*/ 0},
+                       {"mhlo.gather", /*operand*/ 0}};
+
+  const std::unordered_map<std::string, std::set<int>> kShapeCalcOperandMap = {
+      {"mhlo.real_dynamic_slice",
+       {/*start_indices*/ 1, /*limit_indices*/ 2, /*strides*/ 3}},
+      {"mhlo.dynamic_pad",
+       {/*edge_padding_low*/ 2, /*edge_padding_high*/ 3,
+        /*interior_padding*/ 4}},
+      {"mhlo.dynamic_reshape", {/*shape*/ 1}},
+      {"mhlo.dynamic_iota", {/*shape*/ 0}},
+      {"mhlo.dynamic_broadcast_in_dim", {/*out_dim_size*/ 1}},
+      {"mhlo.dynamic_gather", {/*slice_sizes*/ 2}},
+      {"mhlo.dynamic_conv", {/*paddings*/ 2}},
+      {"mhlo.if", {/*pred*/ 0}},
+      {"mhlo.dynamic_rng_uniform", {/*start*/ 0, /*limit*/ 1, /*shape*/ 2}}};
+
+  // add output OP into marked set if it is a I64 scalar and placment is CPU.
+  void markI64ReturnedCPUScalarOps(FuncOp func,
+                                   DenseMap<Operation*, bool>& marked_ops);
+  // Update marked set.
+  // If a OP is in marked set, add all of its operands to marked set.
+  // Add some operands of dynamic shape OPs into marked set according to lookup
+  // table.
+  void markShapeCalculationOps(FuncOp func,
+                               DenseMap<Operation*, bool>& marked_ops);
+
+  // Get placement vector of func's output.
+  SmallVector<llvm::StringRef, 4> getOutputPlacements(FuncOp main_func);
+
+  // Get Op's placement according to its attr
+  PlacementType getOpPlacement(Operation* op);
+
+  // Get tensor's placement
+  PlacementType getTensorPlacement(Operation* dst, size_t operand_idx);
+
+  // Get input argument's placment.
+  PlacementType getArgumentPlacement(Value arg);
+
+  // Enforce output's placement.
+  void enforceOutputPlacement(
+      Operation* dst, FuncOp main_func,
+      SmallVector<std::pair<Operation*, size_t>, 4>& d2h_worklist,
+      SmallVector<std::pair<Operation*, size_t>, 4>& h2d_worklist);
+
+  // insert H2D or D2H Op.
+  void insertMemcpy(Operation* dst, size_t operand_index, bool is_h2d);
+};
+
+// Place the subgraph that is the producer of any shape operands
+// on Host
+// TODO(disc): handle when TupleOp exists in shape_calc_ops
+void PlaceShapeCalcOnHost::placeShapeCalcSubgraphOnHost() {
+  ModuleOp module = getOperation();
+  Builder builder(&getContext());
+  llvm::DenseMap<Operation*, bool> shape_calc_ops;
+
+  for (FuncOp func : module.getOps<FuncOp>()) {
```

Comment:
Why isn't this a FunctionPass?

Reply:
Actually you're modifying function attributes, which isn't safe in a FunctionPass I think.